nan.h

```diff
@@ -1592,51 +1592,64 @@ class Callback {
 /* abstract */ class AsyncProgressWorker : public AsyncWorker {
  public:
   explicit AsyncProgressWorker(Callback *callback_)
-      : AsyncWorker(callback_), asyncdata_(NULL), asyncsize_(0) {
-    async = new uv_async_t;
-    uv_async_init(
-        uv_default_loop()
-      , async
-      , AsyncProgress_
-    );
-    async->data = this;
-
-    uv_mutex_init(&async_lock);
+      : AsyncWorker(callback_) {
   }
 
   virtual ~AsyncProgressWorker() {
-    uv_mutex_destroy(&async_lock);
-
-    if (asyncdata_) {
-      delete[] asyncdata_;
-    }
   }
 
-  void WorkProgress() {
-    uv_mutex_lock(&async_lock);
-    char *data = asyncdata_;
-    size_t size = asyncsize_;
-    asyncdata_ = NULL;
-    uv_mutex_unlock(&async_lock);
-
+  void WorkProgress(char* data, size_t size) {
     // Dont send progress events after we've already completed.
     if (callback) {
         HandleProgressCallback(data, size);
     }
-    delete[] data;
   }
 
+  typedef struct asyncdata_t {
+      uv_async_t* handle;
+      char* data;
+      size_t size;
+      AsyncProgressWorker* worker;
+  } asyncdata_t;
+
   class ExecutionProgress {
     friend class AsyncProgressWorker;
    public:
     // You could do fancy generics with templates here.
     void Send(const char* data, size_t size) const {
-        that_->SendProgress_(data, size);
+      asyncdata_t* asyncdata = new asyncdata_t;
+      asyncdata->data = new char[size];
+      memcpy(asyncdata->data, data, size);
+      asyncdata->size = size;
+      asyncdata->handle = new uv_async_t;
+      asyncdata->worker = that_;
+      uv_async_init(
```

Comment:
This is unsound.  `uv_async_init()` needs to be called on the event loop thread (the main thread), only `uv_async_send()` is thread-safe.